File: 2024/day-09/day9.py
def process_input():
  with open ("./day-09/input.txt", "r") as f:
    input = f.read().strip()
    return input

def solve_part1():
  files = []
  for id in range(0, len(input)):
    character = id // 2 if id % 2 == 0 else "."
    repeats = int(input[id])
    files.extend([character] * repeats)

  total = 0
  i, j = 0, len(files) -1
  while i < len(files):
    if files[i] != ".":
      i += 1
      continue
    # Avoided modifying the files array and instead updated the index j
    # to avoid re-building the list (at O(n) cost) in each iteration.
    while files[j] == ".":
      files[j] = 0
      j -= 1
    files[i] = files[j]
    files[j] = 0
    j -= 1
    i += 1

  total = sum(c * i for i, c in enumerate(files))
  return total

def solve_part2():
  files = []
  for id in range(0, len(input)):
    character = id // 2 if id % 2 == 0 else "."
    repeats = int(input[id])
    files.extend([character] * repeats)

  dots = [] # each dot = [start idx, end idx, count]
  nums = [] # each num = [start idx, end idx, count, char] 
  
  i = 0
  while i < len(files):
    starting_index = i
    starting_char = files[starting_index]
    while i + 1 < len(files) and files[i + 1] == starting_char:
      i += 1
    
    if starting_char == ".":
      dots.append([starting_index, i, i - starting_index + 1, starting_char])
    else:
      nums.append([starting_index, i, i - starting_index + 1, starting_char])

    i += 1 

  nums.reverse() # reverse nums so we can iterate through more easily.

  for num_grp in nums:
    for dot_grp in dots:
      if num_grp[2] > dot_grp[2]:
        pass
      elif num_grp[2] == dot_grp[2] and num_grp[0] > dot_grp[0]:
        num_grp[0] = dot_grp[0]
        num_grp[1] = dot_grp[1]
        dots.remove(dot_grp)
        break
      elif num_grp[2] < dot_grp[2] and num_grp[0] > dot_grp[0]:
        num_grp[0] = dot_grp[0]
        num_grp[1] = dot_grp[0] + num_grp[2] - 1
        dot_grp[0] += num_grp[2]
        dot_grp[2] = dot_grp[1] - dot_grp[0] + 1
        break
  # Spent forever on the above because i missed out the num_grp[0] > dot_grp[0] condition

  total = 0
  for num_grp in nums:
    for mul in range(num_grp[0], num_grp[1] + 1):
      total += int(num_grp[3]) * mul

  return total

if __name__ == "__main__":
  input = process_input()
  print(solve_part1())
  print(solve_part2())

File: 2024/day-09/day9.go
package main

import (
  "fmt"
  "os"
  "strings"
)

func processInput() []int {
  input, err := os.ReadFile("./day-09/input.txt");
  if err != nil {
    fmt.Println("Error while reading file:", err)
    os.Exit(1)
  }

  trimmedInput := strings.TrimSpace(string(input))

  expanded := []int{}
  for i, ascii := range trimmedInput {
    repeats := int(ascii - '0')
    num := i / 2

    if i % 2 == 0 {
      for k := 0; k < repeats; k++ {
        expanded = append(expanded, num)
      }
    } else {
      for k := 0; k < repeats; k++ {
        expanded = append(expanded, -1) // we use -1 to represent '.'
      }
    }
  }
  return expanded
}

func part1() int {
  parsed := processInput()

  // 2 pointer approach
  total := 0
 r := len(parsed) - 1
  for l, num := range parsed {
    if l > r {
      break // early exit
    }
    if num != -1 {
      total += l * num
    } else {
      for parsed[r] == -1 {
        r -= 1
      }
      total += l * parsed[r]
      r -= 1
    }
  }
  return total
}

type Group struct {
  start int
  end int
  len int
  num int
}

func part2() int {
  parsed := processInput()

  // Preprocess data
  dots := []*Group{} // Store pointer for easier mutation later
  nums := []*Group{}

  i := 0
  for i < len(parsed) {
    start := i
    num := parsed[i]
    // determine the ending index of the current character
    for i + 1 < len(parsed) && parsed[i+1] == num {
      i++
    } 

    if num == -1 {
      group := Group{start: start, end: i, len: i - start + 1, num: num}
      dots = append(dots, &group)
    } else {
      group := Group{start: start, end: i, len: i - start + 1, num: num}
      nums = append(nums, &group)
    }

    i++
  }

  // Reverse nums array
  for i := len(nums)/2-1; i >= 0; i-- {
    opp := len(nums)-1-i
    nums[i], nums[opp] = nums[opp], nums[i]
  }

  for _, numGrp := range nums {
    for _, slots := range dots {
      if numGrp.len > slots.len || numGrp.start < slots.start {
        continue
      }

      if slots.len == numGrp.len {
        numGrp.start = slots.start
        numGrp.end = slots.end
        slots.len = 0 // set length to 0 so slot can no longer be used
        break
      } else {
        numGrp.start = slots.start
        numGrp.end = slots.start + numGrp.len - 1
        slots.start += numGrp.len
        slots.len = slots.end - slots.start + 1
        break
      }
    }
  }

  total := 0
  for _, numGrp := range nums {
    for mult := numGrp.start; mult <= numGrp.end; mult++ {
      total += numGrp.num * mult
    }
  }
  return total
}

func main() {
  fmt.Println("Part 1:", part1())
  fmt.Println("Part 2:", part2())
}
